### PR TITLE
Generate parent-ids / Fix up canDrop logic

### PR DESCRIPTION
**Pull request by asinner**

### CODE REVIEW

File: src/components/FormBuilder.tsx [hunk 1]
```diff
@@ -267,10 +268,15 @@ export class FormBuilder extends React.Component<IFormBuilderProps, {}> {
             <FormBuilderContext>
                 <div className={classnames('form-builder', { 'form-builder-empty': isEmpty })}>
                     {this.props.fields.map(this.renderField)}
+<<<<<<< HEAD
```
Comment (linweigao) [inline src/components/FormBuilder.tsx]: Did you need merge this?

File: docs/src/components/DemoPage.tsx [hunk 2]
```diff
@@ -73,6 +75,35 @@ class DemoPage extends React.Component<void, IState> {
...
+
+        // NOTE(andrews): Prohibit moving non-new fields between parents
+        // (eg. moving non-nested field into nested form)
+        if (source.field.id && source.parentId !== target.parentId) {
+            return false;
+        }
+
+        // NOTE(andrews): Prohibit the field from being dropped on itself.
+        // This prevents the indicator from showing up.
+        if (target.field && source.field.id === target.field.id) {
```
Comment (linweigao) [inline docs/src/components/DemoPage.tsx]: This logic should belong to formBuilder, no?

File: docs/src/components/DemoPage.tsx [hunk 3]
```diff
@@ -73,6 +75,35 @@ class DemoPage extends React.Component<void, IState> {
...
+    private canDrop(source: IDragSourceItem, target: IDropTargetItem): boolean {
+
+        // NOTE(andrews): Prohibit more than one layer of nested fields
+        if (source.field.type === 'NestedForm' &&  target.parentId) {
+            return false;
+        }
+
+        // NOTE(andrews): Prohibit moving non-new fields between parents
+        // (eg. moving non-nested field into nested form)
+        if (source.field.id && source.parentId !== target.parentId) {
```
Comment (linweigao) [inline docs/src/components/DemoPage.tsx]: This logic should belong to nestedFormBuilder or FormBuilder.

File: docs/src/components/DemoPage.tsx [hunk 4]
```diff
@@ -73,6 +75,35 @@ class DemoPage extends React.Component<void, IState> {
...
+        }
+
+        // NOTE(andrews): Prohibit the field from being dropped on itself.
+        // This prevents the indicator from showing up.
+        if (target.field && source.field.id === target.field.id) {
+            return false;
+        }
+
+        // NOTE(andrews): Prohibit the field from dropped back into the same index.
+        if (source.index !== undefined && source.index === target.index - 1) {
```
Comment (linweigao) [inline docs/src/components/DemoPage.tsx]: Sam here

**Review by linweigao — CHANGES_REQUESTED** [review]

File: src/components/FormBuilder.tsx (hunk 1 above)
Comment (asinner) [reply src/components/FormBuilder.tsx]: fixed

File: docs/src/components/DemoPage.tsx (hunk 3 above)
Comment (asinner) [reply docs/src/components/DemoPage.tsx]: done

File: docs/src/components/DemoPage.tsx (hunk 2 above)
Comment (asinner) [reply docs/src/components/DemoPage.tsx]: done

File: docs/src/components/DemoPage.tsx (hunk 4 above)
Comment (asinner) [reply docs/src/components/DemoPage.tsx]: done

File: src/components/NestedForm/NestedFormBuilder.tsx [hunk 5]
```diff
@@ -38,6 +36,21 @@ export class NestedFormBuilder extends React.PureComponent<data.IFieldBuilderPro
         return true;
     }
 
+        // NOTE(andrews) canDrop holds the business logic for determining if the source item can
+    // be dropped on the target item.
+    private canDrop(source: data.IDragSourceItem, target: data.IDropTargetItem): boolean {
+        // NOTE(andrews): Prohibit more than one layer of nested fields
+        if (source.field.type === 'NestedForm' &&  target.parentId) {
```
Comment (linweigao) [inline src/components/NestedForm/NestedFormBuilder.tsx]: source.field.type === this.prop.field.type. Is this better instead of hard coding?

**Review by linweigao — APPROVED** [review]

File: src/components/NestedForm/NestedFormBuilder.tsx (hunk 5 above)
Comment (asinner) [reply src/components/NestedForm/NestedFormBuilder.tsx]: done

**Pull request merged**